partitioning/partitions.go

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package partitioning
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"strconv"
+)
+
+const (
+	defaultNumPartitions = 32
+)
+
+var (
+	// Partition range is expected to provided in the form of startPartition..endPartition.
+	// An example partition range is 0..63.
+	rangeRegexp = regexp.MustCompile(`^([0-9]+)(\.\.([0-9]+))?$`)
+
+	errInvalidPartition = errors.New("invalid partition")
+)
+
+// A PartitionSet is a range of partitions organized as a set.
+type PartitionSet map[uint32]struct{}
+
+// UnmarshalYAML unmarshals YAML into a partition set.
+func (ps *PartitionSet) UnmarshalYAML(f func(interface{}) error) error {
+	*ps = make(PartitionSet, defaultNumPartitions)
+
+	// If YAML contains a single string, attempt to parse out a single range.
+	var s string
+	if err := f(&s); err == nil {
+		return ps.ParseRange(s)
+	}
+
+	// Otherwise try to parse out a list of ranges.
+	var a []interface{}
+	if err := f(&a); err == nil {
+		for _, v := range a {
+			switch c := v.(type) {
+			case string:
+				if err := ps.ParseRange(c); err != nil {
+					return err
+				}
+			case int:
+				ps.Add(uint32(c))
+			default:
+				return fmt.Errorf("unexpected range %v", c)
+			}
+		}
+		return nil
+	}
+
+	// Otherwise try to parse out a single partition.
```

Comment:
Any reason we don't require that everything be of the form `x..y`? A single partition could be represented at `x..x`, assuming the bounds are inclusive. If we do want to allow single integers though perhaps update the comment for `PartitionSet` to reflect that.

Reply:
To support non-contiguous ranges and single partition more naturally. Can update the comment sure thing.